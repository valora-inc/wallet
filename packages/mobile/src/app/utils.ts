import { DappRequestOrigin } from 'src/analytics/types'
import { ActiveDapp } from 'src/app/reducers'

// Assume that if we have an active dapp, any WC or DappKit request comes from the in-app webview.
// Note that this may be incorrect if the user interacts with another dapp while the dapp webview is open.
// I thought about comparing the request url to the active dapp url, but that could fail too if the dapp url we have is different.
// For instance the dapp-list could contain a "example.com", while the request is "dapp.something.com".
export function getDappRequestOrigin(activeDapp: ActiveDapp | null) {
  if (activeDapp) {
    return DappRequestOrigin.InAppWebView
  }

  return DappRequestOrigin.External
}
