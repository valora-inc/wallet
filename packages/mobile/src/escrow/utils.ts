import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'
import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToE164NumberSelector } from 'src/identity/selectors'
import { ContactRecipient } from 'src/recipients/recipient'
import useSelector from 'src/redux/useSelector'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}

export const useEscrowPaymentRecipient = (payment: EscrowedPayment): ContactRecipient => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToE164Number = useSelector(identifierToE164NumberSelector)
  const recipientCache = useSelector((state) => state.recipients.phoneRecipientCache)
  const phoneNumber = identifierToE164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber] ?? {
    name: phoneNumber,
    e164PhoneNumber: phoneNumber,
    contactId: '',
  }

  return recipient
}
