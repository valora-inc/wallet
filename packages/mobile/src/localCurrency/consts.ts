// Supported local currency codes
// Please keep it sorted alphabetically
export enum LocalCurrencyCode {
  AUD = 'AUD',
  BRL = 'BRL',
  CAD = 'CAD',
  COP = 'COP',
  CVE = 'CVE',
  EUR = 'EUR',
  GBP = 'GBP',
  GHS = 'GHS',
  KES = 'KES',
  LRD = 'LRD',
  MXN = 'MXN',
  NGN = 'NGN',
  PHP = 'PHP',
  RUB = 'RUB',
  SLL = 'SLL',
  TRY = 'TRY',
  UAH = 'UAH',
  UGX = 'UGX',
  USD = 'USD',
}

export enum LocalCurrencySymbol {
  AUD = 'A$',
  BRL = 'R$',
  CAD = '$',
  COP = '$',
  CVE = '$',
  EUR = '€',
  GBP = '£',
  GHS = 'GH₵',
  KES = 'KSh',
  LRD = 'L$',
  MXN = '$',
  NGN = '₦',
  PHP = '₱',
  RUB = '₽',
  SLL = 'Le',
  TRY = '₺',
  UAH = '₴',
  UGX = 'USh',
  USD = '$',
}

export const LOCAL_CURRENCY_CODES = Object.values(LocalCurrencyCode)
