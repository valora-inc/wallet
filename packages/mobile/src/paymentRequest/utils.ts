import { hexToBuffer } from '@celo/utils/lib/address'
import { decryptComment, encryptComment } from '@celo/utils/lib/commentEncryption'
import { isE164Number } from '@celo/utils/lib/phoneNumbers'
import { call } from 'redux-saga/effects'
import { MAX_COMMENT_LENGTH } from 'src/config'
import { features } from 'src/flags'
import i18n from 'src/i18n'
import { PaymentRequest, WriteablePaymentRequest } from 'src/paymentRequest/types'
import Logger from 'src/utils/Logger'
import { doFetchDataEncryptionKey } from 'src/web3/dataEncryptionKey'

const TAG = 'paymentRequest/utils'

// Encrypt sensitive data in the payment request using the recipient and sender DEK
export function* encryptPaymentRequest(paymentRequest: WriteablePaymentRequest) {
  Logger.debug(`${TAG}@encryptPaymentRequest`, 'Encrypting payment request')

  const fromKey: Buffer | null = yield call(
    doFetchDataEncryptionKey,
    paymentRequest.requesterAddress
  )
  if (!fromKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No sender key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const toKey: Buffer | null = yield call(doFetchDataEncryptionKey, paymentRequest.requesteeAddress)
  if (!toKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No recipient key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const encryptedPaymentRequest: WriteablePaymentRequest = {
    ...paymentRequest,
  }

  if (paymentRequest.requesterE164Number) {
    // Using the same util as for comment encryption to encrypt the phone number
    // TODO: Consider renaming this util for clarity
    const { comment: encryptedE164Number, success } = encryptComment(
      paymentRequest.requesterE164Number,
      toKey,
      fromKey
    )

    // We intentionally exclude the phone number if we can't encrypt it
    // The request still contains an address which may still map to a contact
    // If the recipient has seen it before.
    encryptedPaymentRequest.requesterE164Number = success ? encryptedE164Number : undefined
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: encryptedComment, success } = encryptComment(comment, toKey, fromKey)
    encryptedPaymentRequest.comment = success ? encryptedComment : comment
  }

  return encryptedPaymentRequest
}

// Decrypt sensitive data in the payment request using the user's DEK
export function decryptPaymentRequest(
  paymentRequest: PaymentRequest,
  dataEncryptionKey: string | null,
  isOutgoingRequest: boolean
) {
  Logger.debug(`${TAG}@decryptPaymentRequest`, 'Decrypting payment request')

  if (!dataEncryptionKey) {
    Logger.error(`${TAG}@decryptPaymentRequest`, 'Missing DEK, should never happen.')
    return paymentRequest
  }
  const dekBuffer = hexToBuffer(dataEncryptionKey)

  const decryptedPaymentRequest: PaymentRequest = {
    ...paymentRequest,
  }

  const requesterE164Number = paymentRequest.requesterE164Number
  if (requesterE164Number) {
    const { comment: decryptedRequesterE164Number, success } = decryptComment(
      requesterE164Number,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.requesterE164Number = decryptedRequesterE164Number
    } else if (isE164Number(requesterE164Number)) {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Decrypting requesterE164Number failed, using raw number'
      )
      decryptedPaymentRequest.requesterE164Number = requesterE164Number
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'requesterE164Number appears to be ciphertext, excluding it'
      )
      decryptedPaymentRequest.requesterE164Number = undefined
    }
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: decryptedComment, success } = decryptComment(
      comment,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.comment = decryptedComment
    } else if (comment.length <= MAX_COMMENT_LENGTH) {
      Logger.warn(`${TAG}@decryptPaymentRequest`, 'Decrypting comment failed, using raw comment')
      decryptedPaymentRequest.comment = comment
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Comment appears to be ciphertext, hiding comment'
      )
      decryptedPaymentRequest.comment = i18n.t('commentUnavailable')
    }
  }

  return decryptedPaymentRequest
}

// For cases when the request can't be encrypted, remove sensitive PII
function sanitizePaymentRequest(paymentRequest: WriteablePaymentRequest): WriteablePaymentRequest {
  return {
    ...paymentRequest,
    requesterE164Number: undefined,
  }
}
