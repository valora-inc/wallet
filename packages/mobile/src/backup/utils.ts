import { MnemonicLanguages, normalizeMnemonic } from '@celo/utils/lib/account'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import { useDispatch, useSelector } from 'react-redux'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

export const DAYS_TO_BACKUP = 1
export const DAYS_TO_DELAY = 1 / 24 // 1 hour delay
export const MNEMONIC_SPLITTER = 'celo'

export const MNEMONIC_STORAGE_KEY = 'mnemonic'

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Account Key',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function formatBackupPhraseOnSubmit(phrase: string) {
  return normalizeMnemonic(phrase)
}

function isValidMnemonic(phrase: string, length: number) {
  return !!phrase && countMnemonicWords(formatBackupPhraseOnEdit(phrase)) === length
}

export function isValidBackupPhrase(phrase: string) {
  return isValidMnemonic(phrase, 24)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
