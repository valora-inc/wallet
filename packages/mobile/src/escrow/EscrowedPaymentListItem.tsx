import RequestMessagingCard from '@celo/react-components/components/RequestMessagingCard'
import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { Share, StyleSheet, View } from 'react-native'
import { HomeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import ContactCircle from 'src/components/ContactCircle'
import TokenDisplay from 'src/components/TokenDisplay'
import { EscrowedPayment } from 'src/escrow/actions'
import { useEscrowPaymentRecipient } from 'src/escrow/utils'
import { NotificationBannerCTATypes, NotificationBannerTypes } from 'src/home/NotificationBox'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { useTokenInfo } from 'src/tokens/hooks'
import { divideByWei } from 'src/utils/formatting'
import Logger from 'src/utils/Logger'

interface Props {
  payment: EscrowedPayment
}

const TAG = 'EscrowedPaymentListItem'

function EscrowedPaymentListItem({ payment }: Props) {
  const { t } = useTranslation()
  const recipient = useEscrowPaymentRecipient(payment)
  const tokenInfo = useTokenInfo(payment.tokenAddress)

  const onRemind = async () => {
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.escrow_tx_pending,
      selectedAction: NotificationBannerCTATypes.remind,
    })

    try {
      await Share.share({
        message: t('escrowedPaymentReminderSmsNoData', {
          currency: tokenInfo?.symbol,
        }),
      })
    } catch (error) {
      Logger.error(TAG, 'Error sending reminder', error)
    }
  }

  const onReclaimPayment = () => {
    const reclaimPaymentInput = payment
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.escrow_tx_pending,
      selectedAction: NotificationBannerCTATypes.reclaim,
    })
    navigate(Screens.ReclaimPaymentConfirmationScreen, { reclaimPaymentInput })
  }

  const getCTA = () => {
    const ctas = []
    if (recipient.e164PhoneNumber) {
      ctas.push({
        text: t('remind'),
        onPress: onRemind,
      })
    }
    ctas.push({
      text: t('reclaim'),
      onPress: onReclaimPayment,
    })
    return ctas
  }

  const nameToShow = recipient.name ?? t('unknown')

  return (
    <View style={styles.container}>
      <RequestMessagingCard
        title={t('escrowPaymentNotificationTitle', { mobile: nameToShow })}
        amount={
          <TokenDisplay
            amount={divideByWei(payment.amount)}
            tokenAddress={payment.tokenAddress}
            testID="EscrowedPaymentListItem/amount"
          />
        }
        details={payment.message}
        icon={<ContactCircle recipient={recipient} />}
        callToActions={getCTA()}
        testID={'EscrowedPaymentListItem'}
      />
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
  },
})

export default EscrowedPaymentListItem
