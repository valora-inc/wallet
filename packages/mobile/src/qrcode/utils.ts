import * as RNFS from 'react-native-fs'
import Share from 'react-native-share'
import { call, put, select } from 'redux-saga/effects'
import { showError, showMessage } from 'src/alert/actions'
import { SendEvents } from 'src/analytics/Events'
import { SendOrigin, WalletConnectPairingOrigin } from 'src/analytics/types'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { walletConnectEnabledSelector } from 'src/app/selectors'
import { validateRecipientAddressSuccess } from 'src/identity/actions'
import { E164NumberToAddressType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { UriData, uriDataFromUrl } from 'src/qrcode/schema'
import {
  getRecipientFromAddress,
  recipientHasNumber,
  RecipientInfo,
} from 'src/recipients/recipient'
import { QrCode, SVG } from 'src/send/actions'
import { TransactionDataInput } from 'src/send/SendAmount'
import { handleSendPaymentData } from 'src/send/utils'
import Logger from 'src/utils/Logger'
import { initialiseWalletConnect } from 'src/walletConnect/saga'

export enum BarcodeTypes {
  QR_CODE = 'QR_CODE',
}

const TAG = 'QR/utils'

const QRFileName = '/celo-qr.png'

export async function shareSVGImage(svg: SVG) {
  if (!svg) {
    return
  }
  const data = await new Promise<string>((resolve, reject) => {
    svg.toDataURL((dataURL: string | undefined) => {
      if (dataURL) {
        resolve(dataURL)
      } else {
        // Not supposed to happen, but throw in case it does :)
        reject(new Error('Got invalid SVG data'))
      }
    })
  })

  const path = RNFS.DocumentDirectoryPath + QRFileName
  await RNFS.writeFile(path, data, 'base64')
  return Share.open({
    url: 'file://' + path,
    type: 'image/png',
    failOnCancel: false, // don't throw if user cancels share
  })
}

function* handleSecureSend(
  address: string,
  e164NumberToAddress: E164NumberToAddressType,
  secureSendTxData: TransactionDataInput,
  requesterAddress?: string
) {
  if (!recipientHasNumber(secureSendTxData.recipient)) {
    throw Error('Invalid recipient type for Secure Send, has no mobile number')
  }

  const userScannedAddress = address.toLowerCase()
  const { e164PhoneNumber } = secureSendTxData.recipient
  const possibleReceivingAddresses = e164NumberToAddress[e164PhoneNumber]
  // This should never happen. Secure Send is triggered when there are
  // multiple addresses for a given phone number
  if (!possibleReceivingAddresses) {
    throw Error("No addresses associated with recipient's phone number")
  }

  // Need to add the requester address to the option set in the event
  // a request is coming from an unverified account
  if (requesterAddress && !possibleReceivingAddresses.includes(requesterAddress)) {
    possibleReceivingAddresses.push(requesterAddress)
  }
  const possibleReceivingAddressesFormatted = possibleReceivingAddresses.map((addr) =>
    addr.toLowerCase()
  )
  if (!possibleReceivingAddressesFormatted.includes(userScannedAddress)) {
    const error = ErrorMessages.QR_FAILED_INVALID_RECIPIENT
    ValoraAnalytics.track(SendEvents.send_secure_incorrect, {
      confirmByScan: true,
      error,
    })
    yield put(showMessage(error))
    return false
  }

  ValoraAnalytics.track(SendEvents.send_secure_complete, { confirmByScan: true })
  yield put(validateRecipientAddressSuccess(e164PhoneNumber, userScannedAddress))
  return true
}

export function* handleBarcode(
  barcode: QrCode,
  e164NumberToAddress: E164NumberToAddressType,
  recipientInfo: RecipientInfo,
  secureSendTxData?: TransactionDataInput,
  isOutgoingPaymentRequest?: boolean,
  requesterAddress?: string
) {
  const walletConnectEnabled: boolean = yield select(walletConnectEnabledSelector)
  if (barcode.data.startsWith('wc:') && walletConnectEnabled) {
    navigate(Screens.WalletConnectLoading, { origin: WalletConnectPairingOrigin.Scan })
    yield call(initialiseWalletConnect, barcode.data, WalletConnectPairingOrigin.Scan)
    return
  }

  let qrData: UriData
  try {
    qrData = uriDataFromUrl(barcode.data)
  } catch (e) {
    yield put(showError(ErrorMessages.QR_FAILED_INVALID_ADDRESS))
    Logger.error(TAG, 'qr scan failed', e)
    return
  }

  if (secureSendTxData) {
    const success: boolean = yield call(
      handleSecureSend,
      qrData.address,
      e164NumberToAddress,
      secureSendTxData,
      requesterAddress
    )
    if (!success) {
      return
    }

    if (isOutgoingPaymentRequest) {
      navigate(Screens.PaymentRequestConfirmation, {
        transactionData: secureSendTxData,
        addressJustValidated: true,
      })
    } else {
      navigate(Screens.SendConfirmation, {
        transactionData: secureSendTxData,
        addressJustValidated: true,
        origin: SendOrigin.AppSendFlow,
      })
    }

    return
  }

  const cachedRecipient = getRecipientFromAddress(qrData.address, recipientInfo)

  yield call(handleSendPaymentData, qrData, cachedRecipient, isOutgoingPaymentRequest, true)
}
