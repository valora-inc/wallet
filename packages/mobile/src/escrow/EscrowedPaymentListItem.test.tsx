import { fireEvent, render } from '@testing-library/react-native'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import { Share } from 'react-native'
import { Provider } from 'react-redux'
import EscrowedPaymentListItem from 'src/escrow/EscrowedPaymentListItem'
import { WEI_PER_TOKEN } from 'src/geth/consts'
import { Currency } from 'src/utils/currencies'
import { amountFromComponent, createMockStore, flushMicrotasksQueue } from 'test/utils'
import { mockEscrowedPayment } from 'test/values'

const store = createMockStore()
Share.share = jest.fn()

describe('EscrowedPaymentReminderNotification', () => {
  it('renders correctly', () => {
    const tree = render(
      <Provider store={store}>
        <EscrowedPaymentListItem
          payment={{
            ...mockEscrowedPayment,
            amount: new BigNumber(10).multipliedBy(WEI_PER_TOKEN),
          }}
        />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
    const component = tree.getByTestId('EscrowedPaymentListItem/amount/value')
    // Local currency exchange rate for cUSD is 1.3
    expect(amountFromComponent(component)).toEqual('$13.30')
  })

  it('renders correctly with cEUR', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <EscrowedPaymentListItem
          payment={{
            ...mockEscrowedPayment,
            amount: new BigNumber(10).multipliedBy(WEI_PER_TOKEN),
            currency: Currency.Euro,
          }}
        />
      </Provider>
    )

    const component = getByTestId('EscrowedPaymentListItem/amount/value')
    // Local currency exchange rate for cEUR is 2
    expect(amountFromComponent(component)).toEqual('$20.00')
  })

  it('opens the share dialog', async () => {
    const contact = render(
      <Provider store={store}>
        <EscrowedPaymentListItem payment={mockEscrowedPayment} />
      </Provider>
    )

    fireEvent.press(
      contact.getByTestId('EscrowedPaymentListItem/CallToActions/global:remind/Button')
    )
    await flushMicrotasksQueue()
    expect(Share.share).toHaveBeenCalled()
  })
})
