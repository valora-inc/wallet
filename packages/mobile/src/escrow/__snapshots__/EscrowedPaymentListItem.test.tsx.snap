// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EscrowedPaymentReminderNotification renders correctly 1`] = `
<View
  style={
    Object {
      "marginBottom": 16,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "padding": 16,
        },
        Object {
          "borderRadius": 8,
        },
        Object {
          "elevation": 12,
          "shadowColor": "rgba(156, 164, 169, 0.4)",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 12,
        },
        Array [
          Object {
            "flex": 1,
            "minHeight": 144,
          },
          Object {},
        ],
      ]
    }
    testID="EscrowedPaymentListItem"
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 16,
          }
        }
      >
        <Text
          numberOfLines={2}
          style={
            Object {
              "color": "#9CA4A9",
              "fontFamily": "Inter-Regular",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
            }
          }
          testID="EscrowedPaymentListItem/Title"
        >
          escrowPaymentNotificationTitle, {"mobile":"John Doe"}
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 24,
              "lineHeight": 28,
              "marginBottom": 2,
            }
          }
          testID="EscrowedPaymentListItem/Amount"
        >
          <Text
            testID="EscrowedPaymentListItem/amount"
          >
            
            â‚±
            13.30
          </Text>
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 14,
              "lineHeight": 18,
            }
          }
          testID="EscrowedPaymentListItem/Details"
        />
      </View>
      <View
        style={Object {}}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                Object {
                  "backgroundColor": "hsl(0, 53%, 93%)",
                  "borderRadius": 20,
                  "height": 40,
                  "width": 40,
                },
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "hsl(0, 67%, 24%)",
                    "fontSize": 20,
                  },
                ]
              }
            >
              J
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
        }
      }
      testID="EscrowedPaymentListItem/CallToActions"
    >
      <View
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="EscrowedPaymentListItem/CallToActions/remind/Button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1AB775",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "fontSize": 14,
                "lineHeight": 16,
                "marginRight": 24,
                "minHeight": 16,
                "minWidth": 48,
              },
            ]
          }
        >
          remind
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="EscrowedPaymentListItem/CallToActions/reclaim/Button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1AB775",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "fontSize": 14,
                "lineHeight": 16,
                "marginRight": 24,
                "minHeight": 16,
                "minWidth": 48,
              },
            ]
          }
        >
          reclaim
        </Text>
      </View>
    </View>
  </View>
</View>
`;
